pkg/test-infra/mysql/operation.go

```diff
@@ -16,114 +16,210 @@ package mysql
+			return err
+		}
+	}
+	if err := o.cli.Delete(context.Background(), o.mysql.Sts); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+	return nil
 }
 
-// URI ...
-func (m *MySQL) URI() string {
-	return fmt.Sprintf("root@tcp(%s.%s.svc:3306)/test", m.Svc.Name, m.Svc.Namespace)
+// GetNodes returns MySQL nodes.
+func (o *Ops) GetNodes() ([]clusterTypes.Node, error) {
+	pod := &corev1.Pod{} // only 1 replica
+	err := o.cli.Get(context.Background(), client.ObjectKey{
+		Namespace: o.mysql.Sts.ObjectMeta.Namespace,
+		Name:      fmt.Sprintf("%s-0", o.mysql.Sts.ObjectMeta.Name),
+	}, pod)
+	if err != nil {
+		return []clusterTypes.Node{}, err
+	}
+
+	return []clusterTypes.Node{{
+		Namespace: pod.ObjectMeta.Namespace,
+		PodName:   pod.ObjectMeta.Name,
+		IP:        pod.Status.PodIP,
+		Component: clusterTypes.MySQL,
+		Port:      util.FindPort(pod.ObjectMeta.Name, string(clusterTypes.MySQL), pod.Spec.Containers),
+	}}, nil
 }
 
-// ApplyMySQL ...
-func (m *Ops) ApplyMySQL(spec *Spec) (*MySQL, error) {
-	toCreate, err := m.renderMySQL(spec)
+// GetClientNodes returns the client nodes.
+func (o *Ops) GetClientNodes() ([]clusterTypes.ClientNode, error) {
+	var clientNodes []clusterTypes.ClientNode
+	ips, err := util.GetNodeIPs(o.cli, o.mysql.Sts.Namespace, o.mysql.Sts.ObjectMeta.Labels)
 	if err != nil {
-		return nil, err
+		return clientNodes, err
+	} else if len(ips) == 0 {
+		return clientNodes, errors.New("k8s node not found")
 	}
-	desiredSts := toCreate.Sts.DeepCopy()
-	_, err = controllerutil.CreateOrUpdate(context.TODO(), m.cli, toCreate.Sts, func() error {
-		toCreate.Sts.Spec.Template = desiredSts.Spec.Template
-		return nil
-	})
-	desiredSvc := toCreate.Svc.DeepCopy()
-	_, err = controllerutil.CreateOrUpdate(context.TODO(), m.cli, toCreate.Svc, func() error {
-		clusterIP := toCreate.Svc.Spec.ClusterIP
-		toCreate.Svc.Spec = desiredSvc.Spec
-		toCreate.Svc.Spec.ClusterIP = clusterIP
-		return nil
-	})
-	return toCreate, nil
+
+	svc, err := util.GetServiceByMeta(o.cli, o.mysql.Svc)
+	if err != nil {
+		return clientNodes, err
+	}
+	port := getMySQLNodePort(svc)
+
+	for _, ip := range ips {
+		clientNodes = append(clientNodes, clusterTypes.ClientNode{
+			Namespace:   svc.ObjectMeta.Namespace,
+			ClusterName: svc.ObjectMeta.Labels["instance"],
+			Component:   clusterTypes.MySQL,
+			IP:          ip,
+			Port:        port,
+		})
+	}
+	return clientNodes, nil
 }
 
-// DeleteMySQL ...
-func (m *Ops) DeleteMySQL(ms *MySQL) error {
-	err := m.cli.Delete(context.TODO(), ms.Sts)
-	if err != nil && !errors.IsNotFound(err) {
+func (o *Ops) applyMySQL() error {
+	if err := util.ApplyObject(o.cli, o.mysql.Svc); err != nil {
+		return err
+	}
+	if err := util.ApplyObject(o.cli, o.mysql.Sts); err != nil {
 		return err
 	}
-	err = m.cli.Delete(context.TODO(), ms.Svc)
-	if err != nil && !errors.IsNotFound(err) {
+
+	if err := o.waitMySQLReady(5 * time.Minute); err != nil {
 		return err
 	}
 	return nil
```

Comment:
This can be simplified to just `return o.waitMySQLReady(5 * time.Minute)`

Reply:
addressed in https://github.com/pingcap/tipocket/pull/220/commits/01a18e6682c11bb4ab2b1cec1eea349304badf81. 👍